### DISCUSSION TITLE
build-storybook broken

### DISCUSSION THREAD

**Discussion by jchi2241** [gatsbyjs/gatsby#36293] [Help] [answered]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Summary

`build-storybook` is broken after updating from gatsby v3 to v4, while remaining on React < 18.
```
Module not found: Error: Can't resolve 'react-dom/client' in '/home/jchi/projects/bifrost/frontend/node_modules/.pnpm/gatsby@4.19.2_4ee4e01ebae4918c70a6a4667a6e1cce/node_modules/gatsby/cache-dir'
```

### Steps to Resolve this Issue

Noticed Gatsby defines a global `HAS_REACT_18` in webpack and references it [here](https://github.com/gatsbyjs/gatsby/blob/85391f615fa543aae2921d8aa21e41b36a472762/packages/gatsby/cache-dir/react-dom-utils.js). 

Fixed it by emulating this in `.storybook/main.js` and defining it.
```
...
plugins: [
    new DefinePlugin({
        // Gatsby depends on this global and requires "react-dom/client" if
        // true, else it requires "react-dom".
        HAS_REACT_18: JSON.stringify(
            semver.major(require("react-dom/package.json").version) >= 18
        ),
    }),
],
...
```

Oddly enough the problem goes away with React 18, so not sure what's happening.

Should we include this in the docs? https://www.gatsbyjs.com/docs/how-to/testing/visual-testing-with-storybook/

**Comment by LekoArts** [comment]
I'd suggest adding it here: https://github.com/prismicio-community/storybook-addon-gatsby

**Comment by natapg** [comment]
same here, and I'm not using [storybook-addon-gatsby](https://github.com/prismicio-community/storybook-addon-gatsby)

**Comment by RealSebbooo** [comment]
I am working with React 17. How do i fix this issue?

**Comment by ElSellero** [comment]
Same issue here. I'm also working with React 17.
Is there any solution for this issue? I wasn't able to define an own plugin in my main.js.

**Comment by natapg** [comment, marked as answer]
I fixed by marking it as external in Storybook's Webpack config

```js
// main.js
module.exports = {
  webpackFinal: (config) => {
    config.externals = ["react-dom/client"]
  }
}
```

**Reply by RealSebbooo** [reply]
It works! Thank you very much

**Reply by TwisterMc** [reply]
Thank you!!

**Comment by graysonhicks** [comment]
The code snippet in the docs should now have an automatic fix for this:

https://github.com/gatsbyjs/gatsby/pull/37222

Added same PR for the addon: 

https://github.com/prismicio-community/storybook-addon-gatsby/pull/9
